### ISSUE TITLE
Deprecate the existing `add_job` function more slowly? or add a wrapper function?

### ISSUE THREAD

**Issue by ps2goat** [graphile/worker#319]
It would be great if you could add a wrapper function for `add_job` that has the `integer` parameters for priority and max attempts, and have that call the current `add_job` function that uses `smallint` types. We started a test migration to `0.14.0-alpha.0`, and postgres is unable to automatically cast/infer that a number is a `smallint`. Even though we're only passing in `0` or `-1` or `2`, it fails because no function has a signature with  priority or max_attempts that are `integer` types.

We are thinking of doing this ourselves, but it would be less painful if the existing library had this functionality. Thanks!

**Comment by benjie** [comment]
Adding a function overload may solve this. Could you try the following, please?

```sql
drop function if exists :GRAPHILE_WORKER_SCHEMA.add_job(text, json, text, timestamptz, smallint, text, smallint, text[], text);
drop function if exists :GRAPHILE_WORKER_SCHEMA.add_job(text, json, text, timestamptz, int, text, int, text[], text);

CREATE FUNCTION :GRAPHILE_WORKER_SCHEMA.add_job(identifier text, payload json DEFAULT NULL::json, queue_name text DEFAULT NULL::text, run_at timestamp with time zone DEFAULT NULL::timestamp with time zone, max_attempts smallint DEFAULT NULL::smallint, job_key text DEFAULT NULL::text, priority smallint DEFAULT NULL::smallint, flags text[] DEFAULT NULL::text[], job_key_mode text DEFAULT 'replace'::text) RETURNS :GRAPHILE_WORKER_SCHEMA.jobs
    LANGUAGE plpgsql
    AS $$
declare
  v_job :GRAPHILE_WORKER_SCHEMA.jobs;
begin
  if (job_key is null or job_key_mode is null or job_key_mode in ('replace', 'preserve_run_at')) then
    select * into v_job
    from :GRAPHILE_WORKER_SCHEMA.add_jobs(
      ARRAY[(
        identifier,
        payload,
        queue_name,
        run_at,
        max_attempts,
        job_key,
        priority,
        flags
      ):::GRAPHILE_WORKER_SCHEMA.job_spec],
      (job_key_mode = 'preserve_run_at')
    )
    limit 1;
    return v_job;
  elsif job_key_mode = 'unsafe_dedupe' then
    -- Ensure all the tasks exist
    insert into :GRAPHILE_WORKER_SCHEMA.tasks (identifier)
    values (add_job.identifier)
    on conflict do nothing;
    -- Ensure all the queues exist
    if add_job.queue_name is not null then
      insert into :GRAPHILE_WORKER_SCHEMA.job_queues (queue_name)
      values (add_job.queue_name)
      on conflict do nothing;
    end if;
    -- Insert job, but if one already exists then do nothing, even if the
    -- existing job has already started (and thus represents an out-of-date
    -- world state). This is dangerous because it means that whatever state
    -- change triggered this add_job may not be acted upon (since it happened
    -- after the existing job started executing, but no further job is being
    -- scheduled), but it is useful in very rare circumstances for
    -- de-duplication. If in doubt, DO NOT USE THIS.
    insert into :GRAPHILE_WORKER_SCHEMA.jobs (
      job_queue_id,
      task_id,
      payload,
      run_at,
      max_attempts,
      key,
      priority,
      flags
    )
      select
        job_queues.id,
        tasks.id,
        coalesce(add_job.payload, '{}'::json),
        coalesce(add_job.run_at, now()),
        coalesce(add_job.max_attempts, 25),
        add_job.job_key,
        coalesce(add_job.priority, 0),
        (
          select jsonb_object_agg(flag, true)
          from unnest(add_job.flags) as item(flag)
        )
      from :GRAPHILE_WORKER_SCHEMA.tasks
      left join :GRAPHILE_WORKER_SCHEMA.job_queues
      on job_queues.queue_name = add_job.queue_name
      where tasks.identifier = add_job.identifier
    on conflict (key)
      -- Bump the updated_at so that there's something to return
      do update set
        revision = jobs.revision + 1,
        updated_at = now()
      returning *
      into v_job;
    return v_job;
  else
    raise exception 'Invalid job_key_mode value, expected ''replace'', ''preserve_run_at'' or ''unsafe_dedupe''.' using errcode = 'GWBKM';
  end if;
end;
$$;

create function :GRAPHILE_WORKER_SCHEMA.add_job(text, json, text, timestamptz, int, text, int, text[], text) RETURNS :GRAPHILE_WORKER_SCHEMA.jobs LANGUAGE sql AS $$
select :GRAPHILE_WORKER_SCHEMA.add_job($1, $2, $3, $4, $5::smallint, $6, $7::smallint, $8, $9);
$$;
```

**Comment by ps2goat** [comment]
Due to how we pass arguments based on name (`priority => 1`), we need the parameter names and defaults included in the overload wrapper.

This worked:
```sql
create function :GRAPHILE_WORKER_SCHEMA.add_job(identifier text, payload json DEFAULT NULL::json, queue_name text DEFAULT NULL::text, run_at timestamp with time zone DEFAULT NULL::timestamp with time zone, max_attempts int DEFAULT NULL::int, job_key text DEFAULT NULL::text, priority int DEFAULT NULL::int, flags text[] DEFAULT NULL::text[], job_key_mode text DEFAULT 'replace'::text) RETURNS :GRAPHILE_WORKER_SCHEMA.jobs LANGUAGE sql AS $$
    select :GRAPHILE_WORKER_SCHEMA.add_job(identifier, payload, queue_name, run_at, max_attempts::smallint, job_key, priority::smallint, flags, job_key_mode);
$$;
```

**Comment by benjie** [comment]
Of course 🤦‍♂️ Thanks!

**Comment by ps2goat** [comment]
I appreciate your quick feedback! Is there an ETA on the release/inclusion of this in the next version?

**Comment by benjie** [comment]
Could you expand on the code that it's breaking for you? Because for me it seems to work fine:

```
[worker] # select graphile_worker.add_job('foo', priority => 7, max_attempts => 2);
┌──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│                                                         add_job                                                          │
├──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┤
│ (3,,foo,{},7,"2023-05-04 08:42:32.038226+01",0,2,,"2023-05-04 08:42:32.038226+01","2023-05-04 08:42:32.038226+01",,,,0,) │
└──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
(1 row)

Time: 1.671 ms
```

I guess you are explicitly stating those values are `int` somewhere? You should be able to change those to `smallint` in your existing code and it should work with both old and new versions of worker I think?

```
[worker] # create function echo(i int) returns int as 'select i;' language sql stable;
CREATE FUNCTION
Time: 9.299 ms
[worker] # select echo(1::smallint);
┌──────┐
│ echo │
├──────┤
│    1 │
└──────┘
(1 row)

Time: 0.970 ms
```

**Comment by benjie** [comment]
My bad; I was testing my 0.13.0 installation 🤦‍♂️ With 0.14.0-alpha.0 I can reproduce the issue:

```
[worker] # select * from graphile_worker.add_job('foo', priority => 7, max_attempts => 2);
ERROR:  42883: function graphile_worker.add_job(unknown, priority => integer, max_attempts => integer) does not exist
LINE 1: select * from graphile_worker.add_job('foo', priority => 7, ...
                      ^
HINT:  No function matches the given name and argument types. You might need to add explicit type casts.
LOCATION:  ParseFuncOrColumn, parse_func.c:629
Time: 0.336 ms
```

Having it not work with inferred values is unpleasant. I'm concerned if we add an override though then the `int` version will be favoured and every `add_job` call will call both `add_job` functions and then `add_jobs`, which is somewhat inefficient! I guess that's up to the user to fix though, by using the right function/casts. There's ambiguity if you don't specify `priority`/`max_attempts` as to which version would be called - which is exactly what Postgres warns us about.

**Comment by benjie** [comment]
So we definitely cannot add this overload, because almost all usage would then break:

```
[worker] # select * from graphile_worker.add_job('foo');
ERROR:  42725: function graphile_worker.add_job(unknown) is not unique
LINE 1: select * from graphile_worker.add_job('foo');
                      ^
HINT:  Could not choose a best candidate function. You might need to add explicit type casts.
LOCATION:  ParseFuncOrColumn, parse_func.c:570
```

That's not viable. Most users don't override priority or max_attempts, so we must focus on them.

So an overload is a no-go. Options:

1. change the types back to `int` and cast internally
2. accept it's a breaking change
3. rename `add_job`

I'm actually erring towards 1 because I don't like that `add_job('foo', priority => 7)` doesn't work. Fortunately this can be achieved with an up migration, and anyone who's already using `smallint` should be compatible with the new API because smallint -> int is an implicit cast.

**Comment by benjie** [comment]
Released in `graphile-worker@0.14.0-rc.0`
